### ISSUE TITLE
Open to contribution: adding `torch.nn.functional.scaled_dot_product_attention` support for more architectures

### ISSUE THREAD

**Issue by fxmarty** [huggingface/transformers#28005]
### Feature request

In [`Transformers 4.36`](https://github.com/huggingface/transformers/releases/tag/v4.36.0), we started adding native support of [torch.nn.functional.scaled_dot_product_attention](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html) (SDPA), enabled by default in Transformers: https://huggingface.co/docs/transformers/perf_infer_gpu_one#flashattention-and-memory-efficient-attention-through-pytorchs-scaleddotproductattention

SDPA allows to dispatch to memory-efficient attention, flash attention on supported GPUs (currently NVIDIA-only), and even on [Intel CPUs](https://pytorch.org/blog/new-features-for-ai/#flash-attention-based-scaled-dot-product-algorithm-for-cpu).

For the record, here's a benchmark on some currently supported models:

**[Training benchmark](https://gist.github.com/fxmarty/7e75cc3942d6974e4849093ebea0a331), run on A100-SXM4-80GB.**

| Model     | Batch size | Sequence length | Time per batch (`"eager"`, s) | Time per batch (`"sdpa"`, s) | **Speedup** | Peak memory (`"eager"`, MB) | Peak memory (`"sdpa"`, MB) | **Memory savings**    |
|-----------|------------|-----------------|-------------------------------|------------------------------|-------------|-----------------------------|----------------------------|-----------------------|
| llama2 7b | 4          | 1024            | 1.065                         | 0.90                         | **19.4%**   | 73878.28                    | 45977.81                   | **60.7%**             |
| llama2 7b | 4          | 2048            | OOM                           | 1.87                         | /           | OOM                         | 78394.58                   | **SDPA does not OOM** |
| llama2 7b | 1          | 2048            | 0.64                          | 0.48                         | **32.0%**   | 55557.01                    | 29795.63                   | **86.4%**             |
| llama2 7b | 1          | 3072            | OOM                           | 0.75                         | /           | OOM                         | 37916.08                   | **SDPA does not OOM** |
| llama2 7b | 1          | 4096            | OOM                           | 1.03                         | /           | OOM                         | 46028.14                   | **SDPA does not OOM** |
| llama2 7b | 2          | 4096            | OOM                           | 2.05                         | /           | OOM                         | 78428.14                   | **SDPA does not OOM** |

**[Inference benchmark](https://gist.github.com/fxmarty/5113e4304fbdd38c9c3702ce44683f6a), run on A100-SXM4-80GB.**

| Model            | Batch size | Prompt length | Num new tokens | Per token latency `"eager"` (ms) | Per token latency `"sdpa"` (ms) | **Speedup** |
|------------------|------------|---------------|----------------|----------------------------------|---------------------------------|-------------|
| llama2 13b       | 1          | 1024          | 1 (prefill)    | 178.66                           | 159.36                          | **12.11%**  |
| llama2 13b       | 1          | 100           | 100            | 40.35                            | 37.62                           | **7.28%**   |
| llama2 13b       | 8          | 100           | 100            | 40.55                            | 38.06                           | **6.53%**   |
| Whisper v3 large | 1          | /             | 62             | 20.05                            | 18.90                           | **6.10%**   |
| Whisper v3 large | 8          | /             | 77             | 25.42                            | 24.77                           | **2.59%**   |
| Whisper v3 large | 16         | /             | 77             | 28.51                            | 26.32                           | **8.34%**   |

Previously, we had a partial support of SDPA in [Optimum BetterTransformer](https://huggingface.co/docs/optimum/bettertransformer/overview) but we are now looking to slowly deprecate it in favor of upstream support of SDPA directly in Transformers.

Here are the architectures for which support has been requested:
- [ ] Codegen (https://github.com/huggingface/optimum/issues/1050)
- [ ] LLAVA (https://github.com/huggingface/optimum/issues/1592)
- [ ] Marian (https://github.com/huggingface/optimum/issues/1142)
- [x] Mistral (https://github.com/huggingface/optimum/issues/1553)
- [ ] LongT5 (https://github.com/huggingface/optimum/issues/1506)
- [ ] ViT (https://github.com/huggingface/optimum/issues/1553)

The integration could take inspiration from https://github.com/huggingface/optimum/blob/main/optimum/bettertransformer/models/decoder_models.py & https://github.com/huggingface/optimum/blob/main/optimum/bettertransformer/models/attention.py

### Motivation

Faster training & inference, lower memory requirement

### Your contribution

I may work on some at some point, but contributions are most welcome.

You should refer to https://github.com/huggingface/transformers/pull/26572 to add the support of SDPA for a model, roughly following these steps:
* Create a `XxxSdpaAttention` class inheriting from `XxxAttention` and implement the attention logic using SDPA
* Use `_prepare_4d_causal_attention_mask_for_sdpa` instead of `_prepare_4d_causal_attention_mask` for SDPA
* Use `_prepare_4d_attention_mask_for_sdpa` instead of `_prepare_4d_attention_mask` for SDPA
* Add `_supports_sdpa = True` to `XxxPreTrainedModel`
* Add `"sdpa"` key to `XXX_ATTENTION_CLASSES` in the model modeling file

**Comment by ENate** [comment]
Hi @fxmarty I can take a look at this issue. Of I can ask questions if necessary. Or has anyone taken it already?

**Comment by davidan5** [comment]
does someone know if longT5 and all T5 models are blocked by bias support in flash attention ?

https://github.com/Dao-AILab/flash-attention/pull/617

**Comment by ENate** [comment]
Hi @davidan5 are you working on the implementation?

**Comment by davidan5** [comment]
@ENate I was trying to understand the status and have an estimation of the code change to see if I can contribute.

**Comment by ENate** [comment]
I see.

**Comment by hackyon** [comment]
I'm interested in taking a look at this for the Mistral model if that's still needed. Otherwise, please let me know if there are any other models that still need some work. Thanks

**Comment by ENate** [comment]
Is LongT5 still open?

**Comment by ArthurZucker** [comment]
Mistral is already covered! LongT5 if it is like T5 and has attention bias that might not be supported

**Comment by hackyon** [comment]
Oh yea, looks like you added support for Mistral/Mixtral last month.

It doesn't seem to be supported for BERT yet (I think someone else is working on FA2 but not SDPA), so I'll take a crack at it. 

@ArthurZucker - Please let me know if you know if someone else is already working on SDPA for BERT and I can look for something else to do. Thanks!

**Comment by ArthurZucker** [comment]
Not sure anyone is working on that but bert is already so small that I doubt it will have a lot of impact on perf!

**Comment by abdulfatir** [comment]
@ArthurZucker for the T5 family of models, attention bias is required, so flash-attention won't work for now but torch SDPA can still use the memory efficient kernel from xformers, right? I did some benchmarking with Chronos models (based on T5 architecture) here (https://github.com/amazon-science/chronos-forecasting/issues/33) and there's a clear speedup when using torch SDPA.

**Comment by fxmarty** [comment]
@abdulfatir That's correct

**Comment by abdulfatir** [comment]
I can open a PR for T5 with SDPA then. Are there specific things that I should know of or a reference that can look at?

**Comment by fxmarty** [comment]
@abdulfatir For sure, some specific things that are good to know:

https://github.com/pytorch/pytorch/issues/108108 (`is_causal=True` may not do what you expect)
https://github.com/pytorch/pytorch/issues/110213 (You need https://github.com/huggingface/transformers/blob/416711c3ea88109cf25a9c5f85b4aeee2cb831b5/src/transformers/modeling_attn_mask_utils.py#L189)

example of a PR: https://github.com/huggingface/transformers/pull/29108

**Comment by ArthurZucker** [comment]
FYI going forward we should rather use https://github.com/huggingface/transformers/blob/416711c3ea88109cf25a9c5f85b4aeee2cb831b5/src/transformers/models/llama/modeling_llama.py#L1058 as it is more self contained, easier to debug and maintain than the many paths in the atnn_mask utils

**Comment by sayakpaul** [comment]
Hey @abdulfatir just wanted to check in if you are still working on dropping a PR to add SDPA support for T5? It would tremendously help accelerating diffusion models that use T5.

**Comment by abdulfatir** [comment]
@sayakpaul sorry, I was on vacation. Will look into this now and maybe open a PR in a couple of days. I didn't know that there were diffusion models using the T5 arch. Pretty cool!

**Comment by abdulfatir** [comment]
@fxmarty @ArthurZucker @sayakpaul I have opened a PR #30375 for T5. I still have a couple of questions due to some tests failing. Let's discuss those on the PR.

**Comment by His-Wardship** [comment]
Out of interest, is there any merit in implementing SDPA/FA2 for the DeBERTa family of models? Parametrically, the models are quite small but have relatively high memory costs due to (among other things) the disentangled attention mechanism - however I'm uncertain whether SDPA/FA2 would have a material effect on these costs. 

There was a PR to refactor the DeBERTa modelling scripts (#22105), which appears to have been set aside for now, and I am unsure whether any potential implementations for DeBERTa were gated by that or waiting for it to avoid clashes. There are two tangentially related PRs (#27734, relating to DeBERTa and #28802, relating to BERT, but potentially usable as a starting point) but both appear to have been stuck awaiting some final review or merging for a month or two now.

I think there's still significant utility in the DeBERTa and other BERT family models given their lightweight size and still-leading functionality for text classification, although of course the limelight is held by larger text-completion/multi-modal models, especially with the rapid release of Phi-3, Llama-3 etc. over the past few weeks. 

While most of DeBERTa's use-cases employ short, batched input strings which diminishes the effect of memory efficient optimizations, nevertheless there are tasks, particularly entailment detection between complex sentences, which can create large input strings and very high memory costs.

**Comment by hackyon** [comment]
@His-Wardship - Thanks for your input. We have finally managed to merge #28802 (SDPA for BERT), so hopefully this will unlock other SDPA implementations for other models. I plan to open up a PR for RoBERTa and do some testing to see how if there are any performance gains there.

Regarding DeBERTa, I'm somewhat skeptical whether or not SDPA has support for its disentangled attention mechanism. I think the disentangled attention mechanism requires some extra operations that may not be supported in SDPA.

**Comment by rezacopol** [comment]
run into https://github.com/InternLM/xtuner/issues/633

but xtuner page says it should work, any idea what I might done wrong?

**Comment by abdulfatir** [comment]
@amyeroberts Why was this issue closed? I believe not everything here has been addressed. Can we reopen please?

**Comment by amyeroberts** [comment]
@abdulfatir The issue was closed automatically when #30555 was merged in because it had `Fixes #28005` in the PR description. 

For any future PRs, writing something like `Address #PR_NUMBER` or `Part of #PR_NUMBER` will avoid this

**Comment by OmarManzoor** [comment]
@amyeroberts Could you recommend a model to work on?

**Comment by amyeroberts** [comment]
@OmarManzoor Feel free to add to any model which tickles your fancy and isn't listed here: https://github.com/huggingface/transformers/blob/main/docs/source/en/perf_infer_gpu_one.md#pytorch-scaled-dot-product-attention

**Comment by OmarManzoor** [comment]
Okay, working on `Albert`.

**Comment by Bocchi-Chan2023** [comment]
Is it possible to support moondream2?
Not a large model, but it is one of the best vision model.

**Comment by amyeroberts** [comment]
Hi @Bocchi-Chan2023, there is an implementation of moondream2 available on the hub: https://huggingface.co/vikhyatk/moondream2

**Comment by avishaiElmakies** [comment]
I will start working on OPT

**Comment by AaronZLT** [comment]
Just a stupid question: we are adding native sdpa support in modeling code, since pytorch has provide the fused sdpa?

**Comment by EFord36** [comment]
Hi,,

Not sure if this is the best place to raise this, but I think it should be fairly straightforward to add support for DinoV2, since DinoV2 is using code copied from ViT, and ViT has since been updated (see [here](https://github.com/huggingface/transformers/blob/65bb28444849976f853063edb958b3ef3dd59d12/src/transformers/models/dinov2/modeling_dinov2.py#L61) for comments about copied code, and https://github.com/huggingface/transformers/pull/30555 for adding SDPA to ViT).

I'm using DinoV2 so would be keen for any speedups!

Should I just flag that here, or worth opening a separate issue?

I'm also not sure whether I should be raising support for BetterTransformer for DinoV2 in Optimum instead/as well, but my impression is that native support here is preferable if straightforward so it's used by default?

**Comment by amyeroberts** [comment]
@EFord36 Regarding adding SDPA, you can create a separate issue to track but we can regard this comment as a request for it to be added :) If you or anyone else in the community would like to open a PR to add this to the model we'd be very happy to review a PR! 

Yes, it'd be preferable to be added in transformers as this would mean SDPA is available outside of optimum usage

**Comment by avishaiElmakies** [comment]
I can handle dinoV2

**Comment by avishaiElmakies** [comment]
I have some free time, I will add sdpa to Speech2text. will use bart

**Comment by michaelshekasta** [comment]
@allenyummy  Have you implement sdpa on debertav3? if not, is it sophisticated?

**Comment by ArthurZucker** [comment]
It should not be super complicated, and a good refactoring is quite needed. Once #22105 is merged, we can easily add it!

**Comment by toneyfox** [comment]
ailab_OmniGen
Phi3Transformer does not support an attention implementation through torch.nn.functional.scaled_dot_product_attention yet. Please request the support for this architecture: https://github.com/huggingface/transformers/issues/28005. If you believe this error is a bug, please open an issue in Transformers GitHub repository and load your model with the argument `attn_implementation="eager"` meanwhile. Example: `model = AutoModel.from_pretrained("openai/whisper-tiny", attn_implementation="eager")` 


 HELP ME。
The boss asked how to solve this problem？
![image](https://github.com/user-attachments/assets/76e328cd-88ad-492b-a20d-c0d40ae8e57a)

**Comment by michaelshekasta** [comment]
> It should not be super complicated, and a good refactoring is quite needed. Once #22105 is merged, we can easily add it!

Okay, do we have something like a roadmap? I understand that we should finish the DeBERTaV2 refactoring before implementing FA2 for DeBERTaV3. Do I understand it correctly

**Comment by ArthurZucker** [comment]
Yep! The refactor is now merged, I'll have a look at #34826!

**Comment by sneha09-hub** [comment]
I am using "unsloth/Pixtral-12B-2409-bnb-4bit"
Today I am facing error while loading the model sayin :
'PixtralVisionModel does not support an attention implementation through torch.nn.functional.scaled_dot_product_attention yet'

**Comment by Gorodecki** [comment]
I am using audio class model Wav2Vec2ForCTC
Error:
ValueError: Wav2Vec2ForCTC does not support Flash Attention 2.0 yet.
or
Wav2Vec2ForCTC does not support an attention implementation through torch.nn.functional.scaled_dot_product_attention yet.

**Comment by RuibaiXu** [comment]
Will GptOssForCausalLM support SDPA？
Error: 
   raise ValueError(
ValueError: GptOssForCausalLM does not support an attention implementation through torch.nn.functional.scaled_dot_product_attention yet. Please request the support for this architecture: https://github.com/huggingface/transformers/issues/28005. If you believe this error is a bug, please open an issue in Transformers GitHub repository and load your model with the argument `attn_implementation="eager"` meanwhile. Example: `model = AutoModel.from_pretrained("openai/whisper-tiny", attn_implementation="eager")`

**Comment by ArthurZucker** [comment]
No! `sdpa` does not allow the usage of attention sink 😉
